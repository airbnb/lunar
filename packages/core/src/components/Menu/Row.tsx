import React from 'react';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** Content to display in the row. */
  children: NonNullable<React.ReactNode>;
  /** Double the padding and spacing. */
  spacious?: boolean;
};

/** A non-interactive row within a menu. */
export class MenuRow extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    spacious: false,
  };

  render() {
    const { cx, children, spacious, styles } = this.props;

    return (
      <li role="none">
        <div className={cx(styles.item, spacious && styles.item_spacious)}>{children}</div>
      </li>
    );
  }
}

export default withStyles(({ unit }) => ({
  item: {
    display: 'block',
    width: '100%',
    padding: unit,
    border: 0,
    textAlign: 'left',
  },

  item_spacious: {
    padding: unit * 2,
  },
}))(MenuRow);
