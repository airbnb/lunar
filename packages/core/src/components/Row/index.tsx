import React from 'react';
import withStyles, { WithStylesProps } from '../../composers/withStyles';

export type Props = {
  /** The contents following the primary contents. */
  after?: React.ReactNode;
  /** The visibility of the row's baseline. */
  baseline?: boolean;
  /** The contents preceding the primary contents. */
  before?: React.ReactNode;
  /** The primary contents to render. */
  children: NonNullable<React.ReactNode>;
  /** Render with reduced vertical padding. */
  compact?: boolean;
  /** Max height of row. */
  maxHeight?: number | string;
  /** Align contents in the middle vertically. */
  middleAlign?: boolean;
  /** Render with vertical padding. */
  spacious?: boolean;
  /** The visibility of the row's topline. */
  topline?: boolean;
  /** To use with text truncation; overflow is hidden. */
  truncated?: boolean;
};

/** A presentational block element for displaying horizontal content in up to 3 columns. */
export class Row extends React.Component<Props & WithStylesProps> {
  static defaultProps = {
    after: null,
    baseline: false,
    before: null,
    compact: false,
    flat: false,
    middleAlign: false,
    topline: false,
    truncated: false,
  };

  render() {
    const {
      cx,
      after,
      baseline,
      before,
      children,
      compact,
      maxHeight,
      middleAlign,
      spacious,
      styles,
      topline,
      truncated,
    } = this.props;

    return (
      <div
        className={cx(
          styles.row,
          { maxHeight },
          compact && styles.row_compact,
          spacious && styles.row_spacious,
          middleAlign && styles.row_middleAlign,
          baseline && styles.row_baseline,
          topline && styles.row_topline,
        )}
      >
        {before && (
          <div className={cx(compact ? styles.before_compact : styles.before)}>{before}</div>
        )}

        <div className={cx(styles.primary, truncated && styles.primary_truncated)}>{children}</div>

        {after && <div className={cx(compact ? styles.after_compact : styles.after)}>{after}</div>}
      </div>
    );
  }
}

export default withStyles(({ ui, unit }) => ({
  row: {
    display: 'flex',
  },

  row_compact: {
    paddingBottom: unit * 1.5,
    paddingTop: unit * 1.5,
  },

  row_spacious: {
    paddingBottom: unit * 3,
    paddingTop: unit * 3,
  },

  row_middleAlign: {
    alignItems: 'center',
  },

  row_baseline: {
    borderBottom: ui.border,
  },

  row_topline: {
    borderTop: ui.border,
  },

  after: {
    paddingLeft: unit * 2,
    flexShrink: 0,
  },

  after_compact: {
    paddingLeft: unit,
  },

  before: {
    paddingRight: unit * 2,
    flexShrink: 0,
  },

  before_compact: {
    paddingRight: unit,
  },

  primary: {
    flex: 1,
    maxWidth: '100%',
  },

  primary_truncated: {
    overflow: 'hidden',
  },
}))(Row);
