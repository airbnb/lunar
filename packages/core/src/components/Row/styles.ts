import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ ui, unit }) => ({
  row: {
    display: 'flex',
  },

  row_compact: {
    paddingBottom: unit * 1.5,
    paddingTop: unit * 1.5,
  },

  row_spacious: {
    paddingBottom: unit * 3,
    paddingTop: unit * 3,
  },

  row_middleAlign: {
    alignItems: 'center',
  },

  row_baseline: {
    borderBottom: ui.border,
  },

  row_topline: {
    borderTop: ui.border,
  },

  after: {
    paddingLeft: unit * 2,
    flexShrink: 0,
  },

  after_compact: {
    paddingLeft: unit,
  },

  before: {
    paddingRight: unit * 2,
    flexShrink: 0,
  },

  before_compact: {
    paddingRight: unit,
  },

  primary: {
    flex: 1,
    maxWidth: '100%',
  },

  primary_truncated: {
    overflow: 'hidden',
  },

  inline: {
    display: 'inline-flex',
    alignSelf: 'initial',
  },
});

export default styleSheet;

export { styleSheet };
