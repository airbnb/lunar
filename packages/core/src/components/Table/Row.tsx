import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import { STATUSES } from '../../constants';

const statusPropType = mutuallyExclusiveTrueProps(...STATUSES);

export type Props = {
  /** Table cells to render. */
  children: NonNullable<React.ReactNode>;
  /** Dangerous/failure row (red). */
  danger?: boolean;
  /** Informational row (blue). */
  info?: boolean;
  /** Muted/disabled row (gray). */
  muted?: boolean;
  /** Notice row. */
  notice?: boolean;
  /** Successful row (green). */
  success?: boolean;
  /** Warning row (yellow). */
  warning?: boolean;
};

/** A table row that contains multiple cells. */
export class TableRow extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    danger: statusPropType,
    info: statusPropType,
    muted: statusPropType,
    notice: statusPropType,
    success: statusPropType,
    warning: statusPropType,
  };

  static defaultProps = {
    danger: false,
    info: false,
    muted: false,
    notice: false,
    success: false,
    warning: false,
  };

  render() {
    const {
      cx,
      children,
      danger,
      info,
      muted,
      notice,
      styles,
      success,
      warning,
      ...props
    } = this.props;

    return (
      <tr
        {...props}
        className={cx(
          danger && styles.row_danger,
          info && styles.row_info,
          muted && styles.row_muted,
          notice && styles.row_notice,
          success && styles.row_success,
          warning && styles.row_warning,
        )}
      >
        {children}
      </tr>
    );
  }
}

export default withStyles(({ color }) => {
  function createRow(hex: string) {
    return {
      // Overrides table specificity
      '@selectors': {
        ':nth-child(n) > td': {
          backgroundColor: hex,
        },
      },
    };
  }

  return {
    row_danger: createRow(color.core.danger[0]),
    row_info: createRow(color.core.primary[0]),
    row_muted: createRow(color.core.neutral[0]),
    row_notice: createRow(color.core.secondary[0]),
    row_success: createRow(color.core.success[0]),
    row_warning: createRow(color.core.warning[0]),
  };
})(TableRow);
