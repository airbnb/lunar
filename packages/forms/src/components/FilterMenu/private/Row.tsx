import React from 'react';
import BaseMenuRow, { MenuRowProps } from '@airbnb/lunar/lib/components/Menu/Row';
import proxyComponent from '@airbnb/lunar/lib/utils/proxyComponent';

export default proxyComponent(BaseMenuRow, ({ children }: MenuRowProps) => (
  <BaseMenuRow spacious>{children}</BaseMenuRow>
));
