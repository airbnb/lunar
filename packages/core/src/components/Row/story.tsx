import React from 'react';
import IconAddAlt from '@airbnb/lunar-icons/lib/interface/IconAddAlt';
import LoremIpsum from ':storybook/components/LoremIpsum';
import Text from '../Text';
import Button from '../Button';
import Row from '.';

export default {
  title: 'Core/Row',
  parameters: {
    inspectComponents: [Row],
  },
};

export function primaryContent() {
  return (
    <Row>
      <Text>This row only has primary content in it.</Text>
    </Row>
  );
}

primaryContent.story = {
  name: 'Primary content.',
};

export function beforeContent() {
  return (
    <Row before={<img src="http://via.placeholder.com/50x50" alt="" />}>
      <Text>This row has a before component.</Text>
    </Row>
  );
}

beforeContent.story = {
  name: 'Before content.',
};

export function afterContent() {
  return (
    <Row after={<Button>Take an action</Button>}>
      <Text>This row has an after component.</Text>
    </Row>
  );
}

afterContent.story = {
  name: 'After content.',
};

export function bothBeforeAndAfterContent() {
  return (
    <Row
      middleAlign
      before={<img src="http://via.placeholder.com/50x50" alt="" />}
      after={<Button>Take an action</Button>}
    >
      <Text>This row has both before and after, and is aligned in the middle vertically.</Text>
    </Row>
  );
}

bothBeforeAndAfterContent.story = {
  name: 'Both before and after content.',
};

export function withInline() {
  return (
    <Row inline middleAlign after={<IconAddAlt decorative />}>
      <Text>Inline row with after content, middle aligned.</Text>
    </Row>
  );
}

withInline.story = {
  name: 'With inline.',
};

export function withLongContent() {
  return (
    <Row
      before={<img src="http://via.placeholder.com/50x50" alt="" />}
      after={<Button>Take an action</Button>}
    >
      <Text>
        This is some very long primary content.
        <LoremIpsum />
      </Text>
    </Row>
  );
}

withLongContent.story = {
  name: 'With long content.',
};

export function withToplineAndBaseline() {
  return (
    <Row topline baseline spacious after={<Button>Take an action</Button>}>
      <Text>A row where both topline and baseline are true.</Text>
    </Row>
  );
}

withToplineAndBaseline.story = {
  name: 'With topline and baseline.',
};

export function allLineOptions() {
  return (
    <>
      <div>
        <Row topline spacious after={<Button>Take an action</Button>}>
          <Text>A row where only topline is true.</Text>
        </Row>

        <Row topline baseline spacious after={<Button>Take an action</Button>}>
          <Text>A row with both topline and baseline.</Text>
        </Row>

        <Row baseline spacious after={<Button>Take an action</Button>}>
          <Text>A row where only baseline is true.</Text>
        </Row>
      </div>
    </>
  );
}

allLineOptions.story = {
  name: 'All line options.',
};

export function allPaddingOptions() {
  return (
    <>
      <div>
        <Row spacious topline after={<Button>Take an action</Button>}>
          <Text>A row with spacious vertical padding (24px).</Text>
        </Row>

        <Row compact baseline topline after={<Button>Take an action</Button>}>
          <Text>A row with compact vertical padding (12px).</Text>
        </Row>

        <Row baseline after={<Button>Take an action</Button>}>
          <Text>A row with no padding.</Text>
        </Row>
      </div>
    </>
  );
}

allPaddingOptions.story = {
  name: 'All padding options.',
};

export function withAMinHeight() {
  return (
    <Row topline baseline minHeight={200}>
      <Text>A row with a min height.</Text>
    </Row>
  );
}

withAMinHeight.story = {
  name: 'With a min height.',
};

export function withAMaxHeight() {
  return (
    <Row topline baseline maxHeight={50}>
      <Text>A row with a max height.</Text>
    </Row>
  );
}

withAMaxHeight.story = {
  name: 'With a max height.',
};
