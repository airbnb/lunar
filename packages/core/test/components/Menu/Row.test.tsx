import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import Row from '../../../src/components/Menu/Row';

describe('<MenuRow />', () => {
  it('renders a list with the correct role', () => {
    const wrapper = shallowWithStyles(<Row>Foo</Row>);

    expect(wrapper.is('li')).toBe(true);
    expect(wrapper.prop('role')).toBe('none');
    expect(wrapper.contains('Foo')).toBe(true);
  });

  it('renders spacious', () => {
    const wrapper = shallowWithStyles(<Row spacious>Foo</Row>);

    expect(wrapper).toMatchSnapshot();
  });
});
