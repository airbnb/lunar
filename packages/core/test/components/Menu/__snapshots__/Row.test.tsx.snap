// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuRow /> renders spacious 1`] = `
<li
  role="none"
>
  <div
    className="item_rr6y3q-o_O-item_spacious_1x0fg6n"
  >
    Foo
  </div>
</li>
`;
